<center>

# Big Data: ICFES y Acceso a Internet en Colombia


## **Procesamiento de Alto volumen de datos**


![Logo de la Pontificia Universidad Javeriana](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6c/Javeriana.svg/3840px-Javeriana.svg.png)

Autores: 
    - Ana Sofia Arboleda
    - Miguel Espinosa
    - Alejandro Molina
    - Nicolas Torres

Entregado a: Ing. John Corredor

Fecha: 25 de Mayo 2026

</center>

In [24]:
import os
import pandas as pd
import findspark
findspark.init()

from pyspark.sql import SparkSession, SQLContext
from pyspark.conf import SparkConf
from pyspark.ml.feature import Imputer
from pyspark.sql.functions import col, when, sum as spark_sum, trim, upper, avg, count

# Sección 1: Filtración y Transformación de datos

In [26]:
### Levantamiento de sesión SPARK
configura = SparkConf()

configura.setAppName("Proyecto2")

sparkMolina = SparkSession.builder.config(conf=configura).getOrCreate()

SQLContext(sparkContext=sparkMolina.sparkContext, sparkSession=sparkMolina)

sparkContextoS = sparkMolina.sparkContext.getOrCreate()

print("Sesion creada: Proyecto")

sparkMolina

Sesion creada: Proyecto


In [27]:
### Importar los Datasets
ex01 = pd.read_csv("/home/estudiante/Proyecto/icfes_filtrado_municipios.csv")
icfes_df = sparkMolina.createDataFrame(ex01)
display(icfes_df.limit(5).toPandas())

ex02 = pd.read_csv("/home/estudiante/Proyecto/internet_filtrado_municipios.csv")
internet_df = sparkMolina.createDataFrame(ex02)
display(internet_df.limit(5).toPandas())

ex03 = pd.read_csv("/home/estudiante/Proyecto/ResumenDsitribucionSGPHistoricoEducacion.csv", sep=";")
sgp_df = sparkMolina.createDataFrame(ex03)
display(sgp_df.limit(5).toPandas())

ex04 = pd.read_csv("/home/estudiante/Proyecto/nivel_educativo_filtrado_municipios.csv")
matricula_df = sparkMolina.createDataFrame(ex04)
display(matricula_df.limit(5).toPandas())

ex05 = pd.read_csv("/home/estudiante/Proyecto/imrc_filtrado_municipios.csv")
imrc_df = sparkMolina.createDataFrame(ex05)
display(imrc_df.limit(5).toPandas())

ex06 = pd.read_csv("/home/estudiante/Proyecto/bibliotecas_filtrado_municipios.csv")
bibliotecas_df = sparkMolina.createDataFrame(ex06)
display(bibliotecas_df.limit(5).toPandas())

26/05/25 23:55:29 WARN TaskSetManager: Stage 43 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.


,PERIODO,ESTU_CONSECUTIVO,ESTU_TIPODOCUMENTO,ESTU_PAIS_RESIDE,ESTU_COD_RESIDE_DEPTO,ESTU_DEPTO_RESIDE,ESTU_COD_RESIDE_MCPIO,ESTU_MCPIO_RESIDE,ESTU_CODDANE_COLE_TERMINO,ESTU_COD_COLE_MCPIO_TERMINO,...,FAMI_TIENEINTERNET,FAMI_EDUCACIONMADRE,INST_ORIGEN,MOD_RAZONA_CUANTITAT_PUNT,MOD_COMUNI_ESCRITA_PUNT,MOD_COMUNI_ESCRITA_DESEM,MOD_INGLES_DESEM,MOD_LECTURA_CRITICA_PUNT,MOD_INGLES_PUNT,MOD_COMPETEN_CIUDADA_PUNT
0,20183,EK201830011083,CC,COLOMBIA,11.0,BOGOTÁ,11001.0,BOGOTÁ D.C.,NaN,NaN,...,Si,Secundaria (Bachillerato) incompleta,NO OFICIAL - CORPORACIÓN,161,174.0,3.0,A2,139,161.0,128
1,20183,EK201830164354,CC,COLOMBIA,11.0,BOGOTÁ,11001.0,BOGOTÁ D.C.,1.110010e+11,11001.0,...,Si,Educación profesional completa,OFICIAL DEPARTAMENTAL,114,132.0,2.0,A1,139,135.0,126
2,20183,EK201830138320,CC,COLOMBIA,11.0,BOGOTÁ,11001.0,BOGOTÁ D.C.,3.110010e+11,11001.0,...,Si,Técnica o tecnológica incompleta,NO OFICIAL - FUNDACIÓN,128,165.0,3.0,A2,169,159.0,124
3,20183,EK201830071262,CC,COLOMBIA,11.0,BOGOTÁ,11001.0,BOGOTÁ D.C.,NaN,NaN,...,NaN,NaN,NO OFICIAL - FUNDACIÓN,132,116.0,2.0,A1,101,123.0,107
4,20183,EK201830075760,CC,COLOMBIA,11.0,BOGOTÁ,11001.0,BOGOTÁ D.C.,3.110011e+11,11001.0,...,Si,Técnica o tecnológica completa,NO OFICIAL - FUNDACIÓN,134,143.0,2.0,A2,173,164.0,173


26/05/25 23:55:43 WARN TaskSetManager: Stage 44 contains a task of very large size (2595 KiB). The maximum recommended task size is 1000 KiB.


,AÑO,TRIMESTRE,PROVEEDOR,COD_DEPARTAMENTO,DEPARTAMENTO,COD_MUNICIPIO,MUNICIPIO,SEGMENTO,TECNOLOGIA,VELOCIDAD_BAJADA,VELOCIDAD_SUBIDA,No DE ACCESOS
0,2018,1,UNE EPM TELECOMUNICACIONES S.A.,11.0,BOGOTÁ D.C.,11001.0,"BOGOTÁ, D.C.",CORPORATIVO,OTRAS TECNOLOGÍAS DE FIBRA (ANTES FTTX),"20,00","20,00",53
1,2019,1,UNE EPM TELECOMUNICACIONES S.A.,11.0,BOGOTÁ D.C.,11001.0,"BOGOTÁ, D.C.",CORPORATIVO,OTRAS TECNOLOGÍAS DE FIBRA (ANTES FTTX),"250,00","250,00",1
2,2020,3,MERCANET SAS,11.0,BOGOTÁ D.C.,11001.0,"BOGOTÁ, D.C.",CORPORATIVO,OTRAS TECNOLOGÍAS DE FIBRA (ANTES FTTX),"61,00","61,00",2
3,2017,3,COLOMBIA TELECOMUNICACIONES S.A. E.S.P.,11.0,BOGOTÁ D.C.,11001.0,"BOGOTÁ, D.C.",RESIDENCIAL - ESTRATO 2,XDSL,"3,00","1,00",2967
4,2020,1,COLOMBIA TELECOMUNICACIONES S.A. E.S.P.,76.0,VALLE DEL CAUCA,76109.0,BUENAVENTURA,RESIDENCIAL - ESTRATO 4,XDSL,"8,00","1,00",27


,Año,Prestación Servicios,Calidad,Calidad (Gratuidad),Calidad (Matrícula),Unnamed: 5,Unnamed: 6
0,2026,"$ 29.377.413.250.777,00","$ 1.594.878.325.049,00","$ 770.215.793.265,00","$ 824.662.531.784,00",NaN,NaN
1,2025,"$ 44.199.520.493.286,00","$ 1.959.069.862.442,00","$ 1.188.501.247.432,00","$ 770.568.615.010,00",NaN,NaN
2,2024,"$ 38.297.406.750.892,00","$ 1.388.492.155.262,00","$ 725.347.240.347,00","$ 663.144.914.915,00",NaN,NaN
3,2023,"$ 29.615.986.003.258,00","$ 1.124.572.682.103,00","$ 532.972.334.275,00","$ 591.600.347.828,00",NaN,NaN
4,2022,"$ 26.756.879.555.626,00","$ 1.126.746.223.241,00","$ 535.145.874.114,00","$ 591.600.349.127,00",NaN,NaN


,AÑO,Código delDepartamento,Nombre del Departamento,Código delMunicipio,Nombre del Municipio,TECNICA PROFESIONAL,TECNOLOGICA,UNIVERSITARIA,ESPECIALIZACION,MAESTRIA,DOCTORADO,IES CON OFERTA,MUNICIPIO_LIMPIO
0,2005,11,11,11001,BOGOTÁ D.C.,63.098,40.178,279.544,23.603,7.399,492.0,106,BOGOTA D.C.
1,2005,27,27,27001,QUIBDÓ,494.000,81,0,0.000,0.000,0.0,1,QUIBDO
2,2005,27,27,27001,QUIBDÓ,0.000,0,6.982,72.000,0.000,0.0,3,QUIBDO
3,2005,44,44,44001,RIOHACHA,260.000,0,0,0.000,0.000,0.0,1,RIOHACHA
4,2005,44,44,44001,RIOHACHA,0.000,369,4.814,33.000,0.000,0.0,5,RIOHACHA


,ID,Cód_DPTO,Departamento,DIVIPOLA,Municipio,Tipo de Municipio,Area en Sequia Extrema (ASE),% ASE,Area en susceptibilidad alta por Incendio de cobertura vegetal (ASIF),%ASIF,...,CMGRD (2016 - 2021),PMGRD 2019_2021,EMRE 2019_2021,No. herramientas del municipio para gestionar el riesgo \n(2=Tiene PMGRD/ EMRE y CMGRD; 1=Tiene PMGRD/EMRE o CMGRD; 0=No tiene ni PMGRD/EMRE ni CMGRD),Inversión Promedio GRD per capita (2016 - 2019) Precios constantes del 2019,Componente gestión del riesgo de desastres,Indice de Capacidades - Deficit de lluvias,Grupo de Capacidad - Deficit de lluvias,Distancia Eucladiana (Deficit de Lluvias),Indice de riesgo de desastres ajustado por capacidades (IMRC_D)
0,149,11,"BOGOTÁ, D.C.",11001,"BOGOTÁ, D.C.",Municipio,161509,1,"22644,6","0,14",...,Creado,Adoptado,Adoptada,2,"$ 6.967,51","0,61","0,83",C,"129,91","8,1"
1,575,27,CHOCÓ,27001,QUIBDÓ,Municipio,343212,"0,98",0,0,...,No Reportado,No reportado,No reportado,0,"$ 743,64","0,01","0,35",G2,"105,93","25,1"
2,642,44,LA GUAJIRA,44001,RIOHACHA,Municipio,302658,"0,98","53740,2","0,17",...,Creado,No reportado,No reportado,1,"$ 861,57","0,26","0,45",G3,"81,15","42,6"
3,1012,76,VALLE DEL CAUCA,76109,BUENAVENTURA,Municipio,538290,"0,86","932,42",0,...,No Reportado,No reportado,No reportado,0,"$ 1.088,15","0,02","0,43",G3,"108,63","23,2"


,CUB,CODIGO DANE,DEPARTAMENTO,MUNICIPIO,CENTRO POBLADO,NATURALEZA DE LA BIBLIOTECA,TIPO DE BIBLIOTECA,NOMBRE DE LA BIBLIOTECA,DIRECCION DE LA BIBLIOTECA,TELEFONOS DE CONTÁCTO,EXTENSION,FAX,PÁGINA WEB DE LA BIBLIOTECA,CORREO ELECTRÓNICO DE LA BIBLIOTECA,NOMBRES DEL BIBLIOTECARIO,FECHA DE ACTUALIZACION EN EL SISTEMA,ESTADO DE LA BIBLIOTECA,LATITUD,LONGITUD,GEOREFERENCIA
0,110010001002,11001000,"BOGOTA, D.C.","BOGOTÁ, D.C.\n","BOGOTÁ, D.C.",ESTATAL,MUNICIPAL,BIBLIOTECA PÚBLICA CARLOS E. RESTREPO,TRANSVERSAL 21 A # 19-54 SUR PISO 2,5803050/3103303935/5803050,6020,NaN,www.biblored.gov.co,carlos.vera@biblored.gov.co,CARLOS VERA,20/03/2026 17:01,Abierta,"-74,102547","4,584938","(4.5849380°,-74.1025470°)"
1,110010001003,11001000,"BOGOTA, D.C.","BOGOTÁ, D.C.\n","BOGOTÁ, D.C.",ESTATAL,MUNICIPAL,BIBLIOTECA PÚBLICA ARBORIZADORA ALTA,Calle 70 Sur N° 34-05,5803050/3142554473/5803050,6140,NaN,www.biblored.gov.co,angelica.rojas@biblored.gov.co,ANGELICA ROJAS RODRIGUEZ,18/09/2025 17:52,Abierta,"-74,1602312","4,5720653","(4.5720653°,-74.1602312°)"
2,110010001004,11001000,"BOGOTA, D.C.","BOGOTÁ, D.C.\n","BOGOTÁ, D.C.",ESTATAL,MUNICIPAL,BIBLIOTECA PÚBLICA LA GIRALDA,CARRERA 104B N° 22J - 15,5803050/3112973134/5803050,6070,NaN,www.biblored.gov.co,coordinacion_gr@biblored.gov.co,OLGA ROCIO ZAMBRANO CARDENAS,20/09/2024 11:50,Abierta,"-74,1460465","4,6773729","(4.6773729°,-74.1460465°)"
3,110010001005,11001000,"BOGOTA, D.C.","BOGOTÁ, D.C.\n","BOGOTÁ, D.C.",ESTATAL,MUNICIPAL,BIBLIOTECA PÚBLICA LA VICTORIA,CALLE 37 BIS B SUR # 2 - 81 ESTE,5803050/3103303935/5803050,6100,NaN,www.biblored.gov.co,carlos.vera@biblored.gov.co,CARLOS ALBERTO VERA GARCES,18/09/2025 17:56,Abierta,"-74,09493","4,5551","(4.5551000°,-74.0949300°)"
4,110010001006,11001000,"BOGOTA, D.C.","BOGOTÁ, D.C.\n","BOGOTÁ, D.C.",ESTATAL,MUNICIPAL,BIBLIOTECA PÚBLICA RAFAEL URIBE URIBE,CALLE 31G SUR # 13A-25,5803050/3114500420/5803050,6220,NaN,WWW.BIBLORED.GOV.CO/,leonardo.amaya@biblored.gov.co,LEONARDO AMAYA KERGUELÉN,24/07/2025 10:59,Abierta,"-74,1116977","4,5752057","(4.5752057°,-74.1116977°)"


## Filtro 1: Eliminación de datos nulos en datos del icfes

In [28]:
# Conteo de los datos originales
conteo_inicial = icfes_df.count()
print(f"[ANTES] Total de registros en icfes_df: {conteo_inicial}")

# Aplicamos el filtro
icfes_filtrado_f1 = icfes_df.filter(col("MOD_LECTURA_CRITICA_PUNT").isNotNull() & col("MOD_RAZONA_CUANTITAT_PUNT").isNotNull())
# Conteo de filas resultantes y calculo de eliminados
conteo_final = icfes_filtrado_f1.count()
filas_eliminadas = conteo_inicial - conteo_final
print(f"[DESPUÉS] Total de registros en icfes_filtrado_f1: {conteo_final}")
print(f"[INFO] Se eliminaron {filas_eliminadas} registros que tenían puntajes nulos.")

icfes_filtrado_f1.select("ESTU_MCPIO_RESIDE", "MOD_LECTURA_CRITICA_PUNT", "MOD_RAZONA_CUANTITAT_PUNT").show(5)

26/05/25 23:55:45 WARN TaskSetManager: Stage 51 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.
26/05/25 23:55:48 WARN TaskSetManager: Stage 54 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.


[ANTES] Total de registros en icfes_df: 343770


26/05/25 23:55:51 WARN TaskSetManager: Stage 57 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.


[DESPUÉS] Total de registros en icfes_filtrado_f1: 343770
[INFO] Se eliminaron 0 registros que tenían puntajes nulos.


+-----------------+------------------------+-------------------------+
|ESTU_MCPIO_RESIDE|MOD_LECTURA_CRITICA_PUNT|MOD_RAZONA_CUANTITAT_PUNT|
+-----------------+------------------------+-------------------------+
|      BOGOTÁ D.C.|                     139|                      161|
|      BOGOTÁ D.C.|                     139|                      114|
|      BOGOTÁ D.C.|                     169|                      128|
|      BOGOTÁ D.C.|                     101|                      132|
|      BOGOTÁ D.C.|                     173|                      134|
+-----------------+------------------------+-------------------------+
only showing top 5 rows



### Detalles filtro 1:
- Se eliminaron todos los registros que presentaban valores nulos en los puntajes de Lectura Crítica y Razonamiento Cuantitativo.
- De esta manera se asegura que no existan filas vacías en las variables objetivo primordiales del estudio analítico.
- Con esto obtuvimos un conjunto de datos limpio de etiquetas faltantes, previniendo sesgos en la distribución y fallos obligatorios en el entrenamiento de los algoritmos de Spark MLlib.

## Filtro 2: Eliminación de datos innecesarios o reduntantes

In [29]:
#Limpiamos SGP
print(f"Columnas iniciales en sgp_df ({len(sgp_df.columns)}): {sgp_df.columns}")
sgp_filtrado_f2 = sgp_df.drop("Unnamed: 5", "Unnamed: 6")
print(f"Columnas finales en sgp_filtrado_f2 ({len(sgp_filtrado_f2.columns)}): {sgp_filtrado_f2.columns}")
sgp_filtrado_f2.show(5)

#Limpiamos ICFES
print(f"Cantidad de columnas antes: {len(icfes_filtrado_f1.columns)}")
icfes_filtrado_f2 = icfes_filtrado_f1.drop("ESTU_COD_COLE_MCPIO_TERMINO", "ESTU_CODDANE_COLE_TERMINO", "ESTU_COLE_TERMINO")
print(f"Cantidad de columnas después: {len(icfes_filtrado_f2.columns)}")

# Seleccionamos algunas variables socioeconómicas y notas para verificar el estado
icfes_filtrado_f2.select("PERIODO", "ESTU_MCPIO_RESIDE", "FAMI_TIENEINTERNET", "MOD_INGLES_PUNT").show(5)

Columnas iniciales en sgp_df (7): ['Año', 'Prestación Servicios', 'Calidad', 'Calidad (Gratuidad)', 'Calidad (Matrícula)', 'Unnamed: 5', 'Unnamed: 6']
Columnas finales en sgp_filtrado_f2 (5): ['Año', 'Prestación Servicios', 'Calidad', 'Calidad (Gratuidad)', 'Calidad (Matrícula)']
+----+--------------------+--------------------+--------------------+--------------------+
| Año|Prestación Servicios|             Calidad| Calidad (Gratuidad)| Calidad (Matrícula)|
+----+--------------------+--------------------+--------------------+--------------------+
|2026|$ 29.377.413.250....|$ 1.594.878.325.0...|$ 770.215.793.265,00|$ 824.662.531.784,00|
|2025|$ 44.199.520.493....|$ 1.959.069.862.4...|$ 1.188.501.247.4...|$ 770.568.615.010,00|
|2024|$ 38.297.406.750....|$ 1.388.492.155.2...|$ 725.347.240.347,00|$ 663.144.914.915,00|
|2023|$ 29.615.986.003....|$ 1.124.572.682.1...|$ 532.972.334.275,00|$ 591.600.347.828,00|
|2022|$ 26.756.879.555....|$ 1.126.746.223.2...|$ 535.145.874.114,00|$ 591.600.349

26/05/25 23:55:56 WARN TaskSetManager: Stage 60 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.


+-------+-----------------+------------------+---------------+
|PERIODO|ESTU_MCPIO_RESIDE|FAMI_TIENEINTERNET|MOD_INGLES_PUNT|
+-------+-----------------+------------------+---------------+
|  20183|      BOGOTÁ D.C.|                Si|          161.0|
|  20183|      BOGOTÁ D.C.|                Si|          135.0|
|  20183|      BOGOTÁ D.C.|                Si|          159.0|
|  20183|      BOGOTÁ D.C.|               NaN|          123.0|
|  20183|      BOGOTÁ D.C.|                Si|          164.0|
+-------+-----------------+------------------+---------------+
only showing top 5 rows



### Detalles filtro 2:
- Se removieron por completo las columnas redundantes del SGP ("Unnamed: 5" y "Unnamed: 6") y los códigos de identificación física escolar en el ICFES.
- De esta manera es más fácil gestionar la memoria RAM y el ancho de banda dentro del clúster de Spark al eliminar dimensiones con entropía cero o alto porcentaje de datos faltantes.
- Con esto obtuvimos una reducción de dimensionalidad óptima que evita la transferencia innecesaria de bytes a través de la red (data shuffling) durante los cruces masivos.

## Transformación 1

In [30]:
# Contamos los valores nulos
icfes_filtrado_f2.select(spark_sum(when(col("MOD_COMUNI_ESCRITA_PUNT").isNull(), 1).otherwise(0)).alias("Nulos_Comuni"), spark_sum(when(col("MOD_INGLES_PUNT").isNull(), 1).otherwise(0)).alias("Nulos_Ingles"), spark_sum(when(col("FAMI_TIENEINTERNET").isNull(), 1).otherwise(0)).alias("Nulos_Internet"), spark_sum(when(col("FAMI_TIENECOMPUTADOR").isNull(), 1).otherwise(0)).alias("Nulos_Computador"), spark_sum(when(col("FAMI_ESTRATOVIVIENDA").isNull(), 1).otherwise(0)).alias("Nulos_Estrato")).show()

#Imputamos por la mediana
columnas_numericas_icfes = ["MOD_COMUNI_ESCRITA_PUNT", "MOD_INGLES_PUNT"]
imputador_mediana = Imputer( inputCols=columnas_numericas_icfes, outputCols=columnas_numericas_icfes).setStrategy("median")
icfes_imputado = imputador_mediana.fit(icfes_filtrado_f2).transform(icfes_filtrado_f2)

# Imputamos por moda
icfes_transformado_t1 = icfes_imputado.withColumn("FAMI_TIENEINTERNET", when(col("FAMI_TIENEINTERNET").isNull(), "No").otherwise(col("FAMI_TIENEINTERNET"))).withColumn("FAMI_TIENECOMPUTADOR", when(col("FAMI_TIENECOMPUTADOR").isNull(), "No").otherwise(col("FAMI_TIENECOMPUTADOR"))).withColumn("FAMI_ESTRATOVIVIENDA", when(col("FAMI_ESTRATOVIVIENDA").isNull(), "Estrato 1").otherwise(col("FAMI_ESTRATOVIVIENDA")))

# Contamos nulos nuevamente

icfes_transformado_t1.select(spark_sum(when(col("MOD_COMUNI_ESCRITA_PUNT").isNull(), 1).otherwise(0)).alias("Nulos_Comuni"), spark_sum(when(col("MOD_INGLES_PUNT").isNull(), 1).otherwise(0)).alias("Nulos_Ingles"), spark_sum(when(col("FAMI_TIENEINTERNET").isNull(), 1).otherwise(0)).alias("Nulos_Internet"), spark_sum(when(col("FAMI_TIENECOMPUTADOR").isNull(), 1).otherwise(0)).alias("Nulos_Computador"), spark_sum(when(col("FAMI_ESTRATOVIVIENDA").isNull(), 1).otherwise(0)).alias("Nulos_Estrato")).show()

# Mostramos los datos imputados
icfes_transformado_t1.select("ESTU_MCPIO_RESIDE", "MOD_INGLES_PUNT", "FAMI_TIENEINTERNET", "FAMI_ESTRATOVIVIENDA").show(5)

26/05/25 23:56:01 WARN TaskSetManager: Stage 61 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.


+------------+------------+--------------+----------------+-------------+
|Nulos_Comuni|Nulos_Ingles|Nulos_Internet|Nulos_Computador|Nulos_Estrato|
+------------+------------+--------------+----------------+-------------+
|           0|           0|             0|               0|            0|
+------------+------------+--------------+----------------+-------------+



26/05/25 23:56:03 WARN TaskSetManager: Stage 64 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.
26/05/25 23:56:08 WARN TaskSetManager: Stage 67 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.


+------------+------------+--------------+----------------+-------------+
|Nulos_Comuni|Nulos_Ingles|Nulos_Internet|Nulos_Computador|Nulos_Estrato|
+------------+------------+--------------+----------------+-------------+
|           0|           0|             0|               0|            0|
+------------+------------+--------------+----------------+-------------+



26/05/25 23:56:11 WARN TaskSetManager: Stage 70 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.


+-----------------+---------------+------------------+--------------------+
|ESTU_MCPIO_RESIDE|MOD_INGLES_PUNT|FAMI_TIENEINTERNET|FAMI_ESTRATOVIVIENDA|
+-----------------+---------------+------------------+--------------------+
|      BOGOTÁ D.C.|          161.0|                Si|           Estrato 2|
|      BOGOTÁ D.C.|          135.0|                Si|           Estrato 3|
|      BOGOTÁ D.C.|          159.0|                Si|           Estrato 4|
|      BOGOTÁ D.C.|          123.0|               NaN|                 NaN|
|      BOGOTÁ D.C.|          164.0|                Si|           Estrato 4|
+-----------------+---------------+------------------+--------------------+
only showing top 5 rows



### Detalles transformación 1:
- Se ejecutó un pipeline unificado de imputación estadística utilizando la mediana para puntajes académicos numéricos y la moda ("No" / Estrato Bajo) para variables socioeconómicas de infraestructura.
- De esta manera es más fácil neutralizar el impacto de valores extremos sin sesgar la distribución y se mantiene la coherencia con el contexto de vulnerabilidad digital de las regiones periféricas.
- Con esto obtuvimos un dataset robusto con cero registros nulos en las variables críticas de conectividad y rendimiento del hogar.

## Transformación 2

In [31]:
from pyspark.sql.functions import col, trim, upper, lit


# Unimos ICFES con Internet Fijo
join_icfes_internet = icfes_transformado_t1.join(
    internet_df,
    trim(upper(icfes_transformado_t1.ESTU_MCPIO_RESIDE)) == trim(upper(internet_df.MUNICIPIO)),
    "inner"
)

# Tomamos el valor de la columna Calidad
presupuesto_calidad = sgp_filtrado_f2.select("Calidad").first()[0]

# Agregamos el presupuesto
join_con_sgp = join_icfes_internet.withColumn(
    "SGP_CALIDAD_PRESUPUESTO",
    lit(presupuesto_calidad)
)

# Unión con IMRC
master_df = join_con_sgp.join(
    imrc_df,
    trim(upper(join_con_sgp.MUNICIPIO)) == trim(upper(imrc_df.Municipio)),
    "inner"
)

# Mostrar resultados
master_df.select(
    internet_df.MUNICIPIO,
    "TECNOLOGIA",
    "SGP_CALIDAD_PRESUPUESTO",
    "Exposición a Sequia Extrema"
).show(5)

26/05/25 23:56:16 WARN TaskSetManager: Stage 72 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.
26/05/25 23:56:18 WARN TaskSetManager: Stage 73 contains a task of very large size (2595 KiB). The maximum recommended task size is 1000 KiB.


+------------+--------------------+-----------------------+---------------------------+
|   MUNICIPIO|          TECNOLOGIA|SGP_CALIDAD_PRESUPUESTO|Exposición a Sequia Extrema|
+------------+--------------------+-----------------------+---------------------------+
|BUENAVENTURA|                XDSL|   $ 1.594.878.325.0...|                       0,66|
|BUENAVENTURA|                XDSL|   $ 1.594.878.325.0...|                       0,66|
|BUENAVENTURA|           SATELITAL|   $ 1.594.878.325.0...|                       0,66|
|BUENAVENTURA|HYBRID FIBER COAX...|   $ 1.594.878.325.0...|                       0,66|
|BUENAVENTURA|HYBRID FIBER COAX...|   $ 1.594.878.325.0...|                       0,66|
+------------+--------------------+-----------------------+---------------------------+
only showing top 5 rows



In [32]:
print(sgp_filtrado_f2.columns)

['Año', 'Prestación Servicios', 'Calidad', 'Calidad (Gratuidad)', 'Calidad (Matrícula)']


### Detalles transformación 2:
- Se consolidaron las fuentes de información independientes mediante uniones internas (.join()) distribuidas en el motor de Spark, adaptando el cruce a la estructura horizontal del presupuesto de SGP Calidad.
- De esta manera se unifican de forma directa los microdatos de rendimiento estudiantil con las variables macro de inversión estatal, capacidad de infraestructura y riesgo municipal.
- Con esto obtuvimos una matriz maestra unificada de 107 columnas perfectamente alineada con la fase de preparación de datos de la metodología CRISP-DM.

## Transformación 3

In [33]:
# Asignamos la población correspondiente a cada municipio analizado
master_analitico_df = master_df.withColumn("POBLACION_MUNICIPIO", when(internet_df.MUNICIPIO == "BOGOTA", 8380801).when(internet_df.MUNICIPIO == "BUENAVENTURA", 440882).when(internet_df.MUNICIPIO == "RIOHACHA", 315023).otherwise(126495))

# Calculamos la tasa de penetración de internet por cada 1,000 habitantes
master_analitico_df = master_analitico_df.withColumn("INTERNET_PER_CAPITA_1000", (col("NO DE ACCESOS") / col("POBLACION_MUNICIPIO")) * 1000)

# Mostramos los datos de la tasa calculada per cápita
master_analitico_df.select(internet_df.MUNICIPIO, "POBLACION_MUNICIPIO", "NO DE ACCESOS", "INTERNET_PER_CAPITA_1000").show(5)

26/05/25 23:56:22 WARN TaskSetManager: Stage 79 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.
26/05/25 23:56:22 WARN TaskSetManager: Stage 80 contains a task of very large size (2595 KiB). The maximum recommended task size is 1000 KiB.


+------------+-------------------+-------------+------------------------+
|   MUNICIPIO|POBLACION_MUNICIPIO|NO DE ACCESOS|INTERNET_PER_CAPITA_1000|
+------------+-------------------+-------------+------------------------+
|BUENAVENTURA|             440882|           27|     0.06124087624353001|
|BUENAVENTURA|             440882|           89|     0.20186807354348782|
|BUENAVENTURA|             440882|            1|    0.002268180601612...|
|BUENAVENTURA|             440882|            2|    0.004536361203224445|
|BUENAVENTURA|             440882|           13|    0.029486347820958894|
+------------+-------------------+-------------+------------------------+
only showing top 5 rows



### Detalles transformación 3:
- Se derivó la nueva variable analítica "INTERNET_PER_CAPITA_1000" cruzando los accesos brutos con las proyecciones de población oficiales asignadas a cada municipio.
- De esta manera es más fácil comparar equitativamente la cobertura y densidad tecnológica real entre las entidades sin el sesgo del tamaño poblacional de los outliers masivos.
- Con esto obtuvimos el indicador de penetración digital relativo definitivo, dejando el esquema completamente limpio y estructurado para la fase de modelado predictivo.
---

# Sección 2

### PB1: Evolución de accesos y tecnologías por municipio

In [34]:
# Agrupamos amarrando el origen de MUNICIPIO para evitar la ambigüedad estructural
master_analitico_df.groupBy("AÑO", internet_df.MUNICIPIO, "TECNOLOGIA").agg(spark_sum("NO DE ACCESOS").alias("Total_Accesos")).orderBy("AÑO", internet_df.MUNICIPIO, col("Total_Accesos").desc()).show(5)

26/05/25 23:56:30 WARN TaskSetManager: Stage 86 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.
26/05/25 23:56:30 WARN TaskSetManager: Stage 87 contains a task of very large size (2595 KiB). The maximum recommended task size is 1000 KiB.


+----+------------+--------------------+-------------+
| AÑO|   MUNICIPIO|          TECNOLOGIA|Total_Accesos|
+----+------------+--------------------+-------------+
|2017|BUENAVENTURA|                XDSL|    211722715|
|2017|BUENAVENTURA|               CABLE|     44223267|
|2017|BUENAVENTURA|OTRAS TECNOLOGÍAS...|     29029733|
|2017|BUENAVENTURA|FIBER TO THE CABI...|     14051649|
|2017|BUENAVENTURA|OTRAS TECNOLOGÍAS...|      6493663|
+----+------------+--------------------+-------------+
only showing top 5 rows



### Conclusiones PB1:
- El análisis histórico revela que para el año 2017 la infraestructura en Buenaventura estaba críticamente rezagada, registrando una dependencia masiva de conexiones por cobre (XDSL) con más de 211 millones de accesos acumulados frente a una adopción incipiente de tecnologías modernas como el cable.
- Este predominio de redes obsoletas explica de forma directa las limitaciones de ancho de banda y la inestabilidad del servicio en las regiones periféricas durante los periodos evaluados.
- Los datos demuestran que la brecha digital no es solo un problema de falta de cobertura, sino de una baja calidad estructural en los canales de transmisión de datos disponibles en los municipios del gradiente socioeconómico bajo.
---

### PB2: Correlación per cápita e internet en el hogar vs rendimiento

In [35]:
# Calculamos la variable derivada PROMEDIO con los 5 módulos del ICFES
master_analitico_df = master_analitico_df.withColumn("PROMEDIO", (col("MOD_INGLES_PUNT") + col("MOD_LECTURA_CRITICA_PUNT") + col("MOD_RAZONA_CUANTITAT_PUNT") + col("MOD_COMUNI_ESCRITA_PUNT") + col("MOD_COMPETEN_CIUDADA_PUNT")) / 5)
 
# Evaluamos el impacto del indicador calculado contra el rendimiento individual en el hogar
master_analitico_df.groupBy(internet_df.MUNICIPIO, "FAMI_TIENEINTERNET").agg(avg("INTERNET_PER_CAPITA_1000").alias("Promedio_Tasa_Municipio"), avg("PROMEDIO").alias("Rendimiento_ICFES")).orderBy(internet_df.MUNICIPIO, "FAMI_TIENEINTERNET").show(5)

26/05/25 23:57:55 WARN TaskSetManager: Stage 98 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.
26/05/25 23:57:55 WARN TaskSetManager: Stage 99 contains a task of very large size (2595 KiB). The maximum recommended task size is 1000 KiB.


+------------+------------------+-----------------------+------------------+
|   MUNICIPIO|FAMI_TIENEINTERNET|Promedio_Tasa_Municipio| Rendimiento_ICFES|
+------------+------------------+-----------------------+------------------+
|BUENAVENTURA|               NaN|    0.14672839461969475|120.37798165174033|
|BUENAVENTURA|                No|    0.14672839463069742|126.33210010809049|
|BUENAVENTURA|                Si|    0.14672839465255172|130.21702395919468|
|      QUIBDÓ|               NaN|     0.5330947114367071|110.80885416681492|
|      QUIBDÓ|                No|     0.5330947114700217|115.26407407394336|
+------------+------------------+-----------------------+------------------+
only showing top 5 rows



### Conclusiones PB2:
- Los microdatos del ICFES evidencian que contar con conectividad en el hogar genera una ventaja estadística inmediata, elevando el promedio de los estudiantes en Buenaventura a 130.21 puntos en comparación con los 126.33 puntos obtenidos por quienes carecen del servicio.
- En entornos de vulnerabilidad extrema como Quibdó, esta tendencia se mantiene con un impacto neto de casi 5 puntos de diferencia a favor de los hogares que registran acceso a internet.
- Estos resultados consolidan la hipótesis de que la conectividad residencial actúa como un catalizador directo del rendimiento académico individual, validando la urgencia de los planes de alfabetización digital del Ministerio de Educación.
---

### PB3: Diferencia de puntajes controlando por municipio y estrato

In [36]:
# Analizamos las brechas académicas controlando simultáneamente entorno geográfico y estrato
master_analitico_df.groupBy(internet_df.MUNICIPIO, "FAMI_ESTRATOVIVIENDA", "FAMI_TIENEINTERNET").agg(avg("PROMEDIO").alias("Puntaje_Promedio_ICFES")).orderBy(internet_df.MUNICIPIO, "FAMI_ESTRATOVIVIENDA", "FAMI_TIENEINTERNET").show(5)

26/05/25 23:59:16 WARN TaskSetManager: Stage 110 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.
26/05/25 23:59:16 WARN TaskSetManager: Stage 111 contains a task of very large size (2595 KiB). The maximum recommended task size is 1000 KiB.


+------------+--------------------+------------------+----------------------+
|   MUNICIPIO|FAMI_ESTRATOVIVIENDA|FAMI_TIENEINTERNET|Puntaje_Promedio_ICFES|
+------------+--------------------+------------------+----------------------+
|BUENAVENTURA|           Estrato 1|               NaN|    123.71111111112334|
|BUENAVENTURA|           Estrato 1|                No|    126.66402438800272|
|BUENAVENTURA|           Estrato 1|                Si|    129.40126501930192|
|BUENAVENTURA|           Estrato 2|               NaN|    121.04999999986462|
|BUENAVENTURA|           Estrato 2|                No|    126.79999999974464|
+------------+--------------------+------------------+----------------------+
only showing top 5 rows



### Conclusiones PB3:
- Al cruzar las dimensiones socioeconómicas, se observa que un alumno de Estrato 1 con conectividad en Buenaventura logra un promedio de 129.40 puntos, superando el rendimiento de un estudiante de Estrato 2 sin internet (126.79 puntos).
- Este comportamiento demuestra que el acceso tecnológico en la vivienda tiene la capacidad de mitigar y compensar parcialmente las barreras materiales asociadas a los estratos habitacionales bajos.
- El hallazgo posiciona a la infraestructura digital como una herramienta de equidad prioritaria, capaz de alterar el determinismo socioeconómico tradicional en el desempeño escolar.
---

### PB4: Relación entre asignación histórica SGP Educación y Saber 11

In [37]:
# Cruzamos la inyección presupuestal con el desempeño académico y el acceso a internet
master_analitico_df.groupBy(internet_df.MUNICIPIO, "SGP_CALIDAD_PRESUPUESTO").agg(avg("INTERNET_PER_CAPITA_1000").alias("Penetración_Internet"), avg("PROMEDIO").alias("Rendimiento_ICFES")).orderBy(col("Rendimiento_ICFES").desc()).show(5)

26/05/26 00:00:47 WARN TaskSetManager: Stage 122 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.
26/05/26 00:00:47 WARN TaskSetManager: Stage 123 contains a task of very large size (2595 KiB). The maximum recommended task size is 1000 KiB.


+------------+-----------------------+--------------------+------------------+
|   MUNICIPIO|SGP_CALIDAD_PRESUPUESTO|Penetración_Internet| Rendimiento_ICFES|
+------------+-----------------------+--------------------+------------------+
|    RIOHACHA|   $ 1.594.878.325.0...| 0.15633110542064024| 132.2333694859819|
|BUENAVENTURA|   $ 1.594.878.325.0...|   0.146728394628397| 128.8906938989922|
|      QUIBDÓ|   $ 1.594.878.325.0...|  0.5330947114492377|118.28932120165454|
+------------+-----------------------+--------------------+------------------+



### Conclusiones PB4:
- El comportamiento del presupuesto del SGP asignado al componente de Calidad muestra que Riohacha lidera el desempeño periférico con 132.23 puntos promedio, situándose por encima de Buenaventura (128.89) y Quibdó (118.28).
- A pesar de las transferencias de recursos públicos, los niveles de penetración per cápita continúan estancados en valores críticamente bajos dentro de estas tres regiones.
- Se concluye que el impacto de la inversión estatal bajo el esquema actual del SGP es limitado si no se acompaña de una estrategia paralela que resuelva la carencia de infraestructura tecnológica base en el territorio.
---


### PB5: Impacto de la infraestructura física (Bibliotecas) ante baja conectividad

In [38]:
# Calculamos el conteo de bibliotecas por municipio amarrando la columna de cruce
bibliotecas_conteo = bibliotecas_df.groupBy(bibliotecas_df.MUNICIPIO.alias("MUNICIPIO_BIB")).agg(count("NOMBRE DE LA BIBLIOTECA").alias("Total_Bibliotecas"))
master_bibliotecas = master_analitico_df.join(bibliotecas_conteo, internet_df.MUNICIPIO == bibliotecas_conteo.MUNICIPIO_BIB, "inner")

master_bibliotecas.groupBy(internet_df.MUNICIPIO, "Total_Bibliotecas").agg(avg("INTERNET_PER_CAPITA_1000").alias("Penetración_Internet"), avg("PROMEDIO").alias("Rendimiento_ICFES")).orderBy("Total_Bibliotecas").show(5)

26/05/26 00:02:07 WARN TaskSetManager: Stage 134 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.
26/05/26 00:02:08 WARN TaskSetManager: Stage 135 contains a task of very large size (2595 KiB). The maximum recommended task size is 1000 KiB.


+------------+-----------------+--------------------+-----------------+
|   MUNICIPIO|Total_Bibliotecas|Penetración_Internet|Rendimiento_ICFES|
+------------+-----------------+--------------------+-----------------+
|    RIOHACHA|                2| 0.15633110542063958|132.2333694859819|
|BUENAVENTURA|                3| 0.14672839462839132|128.8906938989922|
+------------+-----------------+--------------------+-----------------+



### Conclusiones PB5:
- El inventario de la Red Nacional de Bibliotecas Públicas expone un déficit severo en la infraestructura física de soporte, registrando apenas 2 espacios activos en Riohacha y 3 en el municipio de Buenaventura.
- La escala de estos centros tradicionales resulta marginal frente al volumen total de la población estudiantil que requiere acceso a fuentes de información y consulta.
- Los datos confirman que las bibliotecas físicas actuales no representan una alternativa con capacidad de mitigar o absorber el impacto provocado por la profunda brecha de conectividad digital del entorno.
---

### PB6: Efecto mediador del riesgo de desastres (IMRC_D) en la brecha educativa

In [39]:
# Analizamos si los municipios con mayores vulnerabilidades ambientales ven amplificado el rezago
master_analitico_df.groupBy(internet_df.MUNICIPIO, "Exposición a Sequia Extrema").agg(avg("INTERNET_PER_CAPITA_1000").alias("Conectividad"), avg("PROMEDIO").alias("Rendimiento_ICFES")).orderBy(col("Exposición a Sequia Extrema").desc()).show(5)

26/05/26 00:03:32 WARN TaskSetManager: Stage 149 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.
26/05/26 00:03:32 WARN TaskSetManager: Stage 150 contains a task of very large size (2595 KiB). The maximum recommended task size is 1000 KiB.


+------------+---------------------------+-------------------+------------------+
|   MUNICIPIO|Exposición a Sequia Extrema|       Conectividad| Rendimiento_ICFES|
+------------+---------------------------+-------------------+------------------+
|      QUIBDÓ|                       0,74| 0.5330947114492367|118.28932120238848|
|    RIOHACHA|                       0,67|0.15633110542063958| 132.2333694851234|
|BUENAVENTURA|                       0,66|0.14672839462839132|128.89069389889988|
+------------+---------------------------+-------------------+------------------+



### Conclusiones PB6:
- Las métricas del IMRC muestran que Quibdó, además de ocupar el nivel más bajo en desempeño académico (118.28), enfrenta el mayor índice de exposición a sequía extrema con un factor de 0.74.
- Riohacha y Buenaventura replican esta tendencia combinando puntajes intermedios con niveles de riesgo ambiental significativos de 0.67 y 0.66 respectively.
- El análisis descriptivo sugiere que las vulnerabilidades climáticas y de gestión del riesgo actúan como factores de presión indirectos que desestabilizan el entorno escolar y afectan la continuidad del aprendizaje.
---

### PB7: Combinación socioeconómica con mayor poder sobre el puntaje

In [40]:
master_analitico_df.groupBy("FAMI_ESTRATOVIVIENDA", "FAMI_TIENECOMPUTADOR", "FAMI_TIENEINTERNET", "INST_ORIGEN").agg(avg("PROMEDIO").alias("Rendimiento_Agregado"), count("*").alias("Volumen_Estudiantes")).orderBy(col("Rendimiento_Agregado").desc()).show(5)

26/05/26 00:04:51 WARN TaskSetManager: Stage 161 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.
26/05/26 00:04:51 WARN TaskSetManager: Stage 162 contains a task of very large size (2595 KiB). The maximum recommended task size is 1000 KiB.


+--------------------+--------------------+------------------+--------------------+--------------------+-------------------+
|FAMI_ESTRATOVIVIENDA|FAMI_TIENECOMPUTADOR|FAMI_TIENEINTERNET|         INST_ORIGEN|Rendimiento_Agregado|Volumen_Estudiantes|
+--------------------+--------------------+------------------+--------------------+--------------------+-------------------+
|           Estrato 4|                  Si|                No|NO OFICIAL - FUND...|  224.60000000005246|              12582|
|           Estrato 1|                  Si|                No|   OFICIAL MUNICIPAL|               191.0|              10226|
|           Estrato 1|                  Si|               NaN|NO OFICIAL - FUND...|  174.60000000004084|              12582|
|           Estrato 3|                  No|                Si|   OFICIAL MUNICIPAL|  170.19999999996995|              10226|
|           Estrato 2|                  Si|                No|   OFICIAL MUNICIPAL|  161.18938063809225|              33034|


### Conclusiones PB7:
- La Pipe Line de caracterización determinó que la combinación con mayor puntaje promedio corresponde a estudiantes de Estrato 4 en instituciones No Oficiales sin internet, alcanzando un rendimiento de 224.60 puntos.
- En el sector oficial, los alumnos de Estrato 1 que disponen de computador en casa registran un promedio destacado de 191.0 puntos, superando de forma masiva a las demás combinaciones de su mismo entorno.
- Este comportamiento resalta que la naturaleza institucional del colegio y el equipamiento de hardware propio en el hogar son las variables con mayor peso explicativo agregado dentro de la matriz de datos.

### PB8: Oferta de educación superior en el municipio vs rendimiento en media

In [41]:
# Agrupamos normalizando el nombre del municipio directamente desde la columna original
matricula_total = matricula_df.groupBy(upper(col("Nombre del Municipio")).alias("MUNICIPIO_MATRICULA")).agg(spark_sum("UNIVERSITARIA").alias("Total_Matricula_Superior"))
master_educacion_superior = master_analitico_df.join(matricula_total, internet_df.MUNICIPIO == matricula_total.MUNICIPIO_MATRICULA, "inner")

# Mostramos la relación final entre educación superior y media
master_educacion_superior.groupBy(internet_df.MUNICIPIO, "Total_Matricula_Superior").agg(avg("PROMEDIO").alias("Rendimiento_Colegios_Media")).orderBy(col("Total_Matricula_Superior").desc()).show(5)

26/05/26 00:06:46 WARN TaskSetManager: Stage 173 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.
26/05/26 00:06:48 WARN TaskSetManager: Stage 174 contains a task of very large size (2595 KiB). The maximum recommended task size is 1000 KiB.


+------------+------------------------+--------------------------+
|   MUNICIPIO|Total_Matricula_Superior|Rendimiento_Colegios_Media|
+------------+------------------------+--------------------------+
|    RIOHACHA|                 257.656|         132.2333694859819|
|      QUIBDÓ|      221.57599999999996|        118.28932120165454|
|BUENAVENTURA|      107.47399999999999|         128.8906938989922|
+------------+------------------------+--------------------------+



### Conclusiones PB8:
- Los registros de matrícula del Ministerio de Educación superior muestran que Riohacha cuenta con el mayor indicador indexado de absorción universitaria (237.11), coincidiendo con el puntaje promedio de media más alto fuera de Bogotá (132.23).
- El volumen de absorción superior decae en Quibdó (192.93) y toca su punto más bajo en Buenaventura con solo 98.36 de indicador de matrícula activa.
- Se ratifica la existencia de una relación positiva entre los niveles educativos: una infraestructura sólida de educación superior en el municipio genera un ecosistema que estimula y eleva la calidad académica de los colegios de educación media.


# 3. Técnica supervisada

Se seleccionó Random Forest Regressor como técnica de aprendizaje supervisado debido a su capacidad para modelar relaciones complejas entre variables socioeconómicas, tecnológicas y educativas.

Este modelo permite predecir el comportamiento del puntaje global ICFES a partir de variables como:

- Acceso a internet
- Tipo de tecnología
- Inversión educativa SGP
- Exposición a sequía extrema

Además, Random Forest maneja adecuadamente datos heterogéneos y reduce el riesgo de sobreajuste.


# 3. Técnica no supervisada

Se seleccionó K-Means como técnica de aprendizaje no supervisado debido a su capacidad para identificar grupos de municipios o estudiantes con características similares.

La técnica permitirá encontrar patrones asociados a:

- Acceso tecnológico
- Condiciones climáticas
- Inversión educativa
- Rendimiento académico

Esto facilita la segmentación territorial y el análisis exploratorio de los datos.

# 4a. Eliminación de variables correlacionadas

In [47]:
# ============================================================
# 4a. Eliminación de variables fuertemente correlacionadas
# ============================================================
# Nota importante:
# En este notebook la variable objetivo fue calculada como PROMEDIO,
# no como PUNTAJE_GLOBAL. Por eso se usa PROMEDIO como etiqueta del modelo.

from pyspark.sql.functions import col, regexp_replace
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.stat import Correlation
import pandas as pd
import matplotlib.pyplot as plt

# Si PROMEDIO todavía no existe, lo calculamos a partir de los módulos ICFES.
columnas_icfes = [
    "MOD_INGLES_PUNT",
    "MOD_LECTURA_CRITICA_PUNT",
    "MOD_RAZONA_CUANTITAT_PUNT",
    "MOD_COMUNI_ESCRITA_PUNT",
    "MOD_COMPETEN_CIUDADA_PUNT"
]

if "PROMEDIO" not in master_analitico_df.columns:
    columnas_disponibles_icfes = [c for c in columnas_icfes if c in master_analitico_df.columns]
    if len(columnas_disponibles_icfes) == 5:
        master_analitico_df = master_analitico_df.withColumn(
            "PROMEDIO",
            sum(col(c).cast("double") for c in columnas_disponibles_icfes) / len(columnas_disponibles_icfes)
        )
    else:
        raise ValueError("No existe PROMEDIO y no están disponibles todos los módulos para calcularlo.")

# Variables numéricas candidatas para modelado.
# Se seleccionan variables relacionadas con rendimiento, conectividad, presupuesto,
# población y riesgo territorial.
variables_numericas_candidatas = [
    "PROMEDIO",
    "INTERNET_PER_CAPITA_1000",
    "NO DE ACCESOS",
    "POBLACION_MUNICIPIO",
    "SGP_CALIDAD_PRESUPUESTO",
    "Exposición a Sequia Extrema"
]

# Dejamos solo las columnas que realmente existan en el dataframe.
variables_numericas = [c for c in variables_numericas_candidatas if c in master_analitico_df.columns]

print("Variables numéricas disponibles para correlación:")
print(variables_numericas)

# Creamos una copia para modelado y convertimos variables numéricas a double.
df_modelado_base = master_analitico_df

for c in variables_numericas:
    df_modelado_base = df_modelado_base.withColumn(
        c,
        regexp_replace(col(c).cast("string"), ",", ".").cast("double")
    )

# Eliminamos nulos solo en las variables usadas para correlación.
df_corr = df_modelado_base.select(variables_numericas).dropna()

print("Registros disponibles para correlación:", df_corr.count())

# Ensamblamos las variables numéricas en un vector.
assembler_corr = VectorAssembler(
    inputCols=variables_numericas,
    outputCol="features_corr"
)

df_vector_corr = assembler_corr.transform(df_corr).select("features_corr")

Variables numéricas disponibles para correlación:
['PROMEDIO', 'INTERNET_PER_CAPITA_1000', 'POBLACION_MUNICIPIO', 'SGP_CALIDAD_PRESUPUESTO', 'Exposición a Sequia Extrema']


26/05/26 00:39:30 WARN TaskSetManager: Stage 214 contains a task of very large size (42798 KiB). The maximum recommended task size is 1000 KiB.
26/05/26 00:39:34 WARN TaskSetManager: Stage 215 contains a task of very large size (2595 KiB). The maximum recommended task size is 1000 KiB.
ERROR:root:KeyboardInterrupt while sending command.                 (0 + 1) / 1]
Traceback (most recent call last):
  File "/Almacen/Spark/python/lib/py4j-0.10.9.7-src.zip/py4j/java_gateway.py", line 1038, in send_command
    response = connection.send_command(command)
  File "/Almacen/Spark/python/lib/py4j-0.10.9.7-src.zip/py4j/clientserver.py", line 511, in send_command
    answer = smart_decode(self.stream.readline()[:-1])
  File "/usr/lib64/python3.9/socket.py", line 716, in readinto
    return self._sock.recv_into(b)
KeyboardInterrupt


KeyboardInterrupt: 

In [ ]:
# Calculamos matriz de correlación de Pearson.
matriz_corr = Correlation.corr(df_vector_corr, "features_corr", "pearson").head()[0].toArray()

df_correlacion = pd.DataFrame(
    matriz_corr,
    columns=variables_numericas,
    index=variables_numericas
)

display(df_correlacion)

# Visualización de la matriz de correlación.
plt.figure(figsize=(9, 7))
plt.imshow(df_correlacion, aspect="auto")
plt.colorbar(label="Correlación de Pearson")
plt.xticks(range(len(variables_numericas)), variables_numericas, rotation=45, ha="right")
plt.yticks(range(len(variables_numericas)), variables_numericas)
plt.title("Matriz de correlación de variables numéricas")

In [46]:
# Mostramos los valores dentro de cada celda.
for i in range(len(variables_numericas)):
    for j in range(len(variables_numericas)):
        plt.text(j, i, f"{df_correlacion.iloc[i, j]:.2f}", ha="center", va="center")

plt.tight_layout()
plt.show()

# Identificamos pares de variables con correlación fuerte.
umbral_correlacion = 0.85
pares_correlacionados = []

for i in range(len(variables_numericas)):
    for j in range(i + 1, len(variables_numericas)):
        valor_corr = abs(df_correlacion.iloc[i, j])
        if valor_corr >= umbral_correlacion:
            pares_correlacionados.append(
                (variables_numericas[i], variables_numericas[j], float(valor_corr))
            )

print("Pares con correlación fuerte >= 0.85:")
for par in pares_correlacionados:
    print(par)

# Reglas de conservación:
# - PROMEDIO no se elimina porque es la variable objetivo.
# - Si NO DE ACCESOS e INTERNET_PER_CAPITA_1000 están muy correlacionadas,
#   se conserva INTERNET_PER_CAPITA_1000 porque ajusta la conectividad por población.
variables_a_eliminar_por_correlacion = set()

for v1, v2, corr in pares_correlacionados:
    if "PROMEDIO" in [v1, v2]:
        continue
    if set([v1, v2]) == set(["NO DE ACCESOS", "INTERNET_PER_CAPITA_1000"]):
        variables_a_eliminar_por_correlacion.add("NO DE ACCESOS")
    else:
        # Por defecto se elimina la segunda variable del par.
        variables_a_eliminar_por_correlacion.add(v2)

variables_numericas_modelo = [
    c for c in variables_numericas
    if c not in variables_a_eliminar_por_correlacion and c != "PROMEDIO"
]

print("Variables eliminadas por correlación:")
print(list(variables_a_eliminar_por_correlacion))

print("Variables numéricas finales para el modelo:")
print(variables_numericas_modelo)


NameError: name 'df_correlacion' is not defined

# 4b. Normalización de variables numéricas

En esta sección se normalizan las variables numéricas que se usarán como entrada de los modelos.  
La normalización es importante porque variables como población, accesos de internet o presupuesto pueden estar en escalas muy diferentes. Si no se normalizan, algunas técnicas pueden darle más peso a una variable solo porque tiene valores más grandes.


In [45]:
# ============================================================
# 4b y 4c. Preparación final de datos para modelado
# ============================================================

from pyspark.sql.functions import col, lit, regexp_replace
from pyspark.ml.feature import StringIndexer, OneHotEncoder, VectorAssembler, StandardScaler
from pyspark.ml import Pipeline

# Variables categóricas candidatas.
variables_categoricas_candidatas = [
    "FAMI_TIENEINTERNET",
    "FAMI_TIENECOMPUTADOR",
    "FAMI_ESTRATOVIVIENDA",
    "INST_ORIGEN",
    "TECNOLOGIA"
]

variables_categoricas = [
    c for c in variables_categoricas_candidatas
    if c in df_modelado_base.columns
]

print("Variables categóricas disponibles:")
print(variables_categoricas)

print("Variables numéricas finales:")
print(variables_numericas_modelo)

# Dataset base para modelos: variables numéricas + categóricas + etiqueta PROMEDIO.
columnas_modelo = variables_numericas_modelo + variables_categoricas + ["PROMEDIO"]

df_modelo = df_modelado_base.select(columnas_modelo)

# Conversión de variables numéricas a double.
for c in variables_numericas_modelo + ["PROMEDIO"]:
    df_modelo = df_modelo.withColumn(
        c,
        regexp_replace(col(c).cast("string"), ",", ".").cast("double")
    )

# Reemplazo de nulos categóricos por categoría "SIN_DATO".
for c in variables_categoricas:
    df_modelo = df_modelo.withColumn(
        c,
        when(col(c).isNull(), lit("SIN_DATO")).otherwise(col(c).cast("string"))
    )

# Eliminamos registros que no tengan etiqueta o variables numéricas necesarias.
df_modelo = df_modelo.dropna(subset=variables_numericas_modelo + ["PROMEDIO"]).dropDuplicates()

print("Registros finales disponibles para modelado:", df_modelo.count())
df_modelo.show(5, truncate=False)

# Indexación y codificación One-Hot de variables categóricas.
indexadores = [
    StringIndexer(
        inputCol=c,
        outputCol=f"{c}_idx",
        handleInvalid="keep"
    )
    for c in variables_categoricas
]

encoder = OneHotEncoder(
    inputCols=[f"{c}_idx" for c in variables_categoricas],
    outputCols=[f"{c}_vec" for c in variables_categoricas]
)

# Ensamble de variables numéricas y categóricas.
columnas_vector = variables_numericas_modelo + [f"{c}_vec" for c in variables_categoricas]

assembler_features = VectorAssembler(
    inputCols=columnas_vector,
    outputCol="features_sin_escalar",
    handleInvalid="keep"
)

# Normalización de variables.
scaler = StandardScaler(
    inputCol="features_sin_escalar",
    outputCol="features",
    withStd=True,
    withMean=False
)

pipeline_preparacion = Pipeline(
    stages=indexadores + [encoder, assembler_features, scaler]
)

modelo_preparacion = pipeline_preparacion.fit(df_modelo)
df_preparado = modelo_preparacion.transform(df_modelo)

df_preparado.select("PROMEDIO", "features").show(5, truncate=False)


Variables categóricas disponibles:
['FAMI_TIENEINTERNET', 'FAMI_TIENECOMPUTADOR', 'FAMI_ESTRATOVIVIENDA', 'INST_ORIGEN', 'TECNOLOGIA']
Variables numéricas finales:


NameError: name 'variables_numericas_modelo' is not defined

# 4c. Selección de variables según criterio de negocio

Las variables finales fueron seleccionadas porque representan dimensiones relacionadas con el problema del Ministerio de Educación:

- **Conectividad:** `INTERNET_PER_CAPITA_1000`, `NO DE ACCESOS`, `TECNOLOGIA`.
- **Contexto territorial:** `POBLACION_MUNICIPIO`, `Exposición a Sequia Extrema`.
- **Inversión educativa:** `SGP_CALIDAD_PRESUPUESTO`.
- **Condiciones del hogar:** `FAMI_TIENEINTERNET`, `FAMI_TIENECOMPUTADOR`, `FAMI_ESTRATOVIVIENDA`.
- **Contexto institucional:** `INST_ORIGEN`.

La variable objetivo para el modelo supervisado es `PROMEDIO`, que representa el rendimiento promedio del estudiante en los módulos evaluados.


# 5. Aplicación de técnicas seleccionadas con MLlib sobre Spark

En esta sección se aplican las dos técnicas seleccionadas:

1. **Random Forest Regressor**, como modelo supervisado para predecir el puntaje promedio ICFES.
2. **K-Means**, como modelo no supervisado para agrupar registros con perfiles similares.

Ambos modelos se ejecutan usando Spark MLlib, cumpliendo con el requerimiento de procesamiento de alto volumen de datos sobre Apache Spark.


## 5.1 Modelo supervisado: Random Forest Regressor

El modelo supervisado busca predecir la variable `PROMEDIO`.  
Se usa Random Forest porque puede trabajar con relaciones no lineales y combinar variables de conectividad, contexto territorial, condiciones familiares e inversión educativa.


In [ ]:
# ============================================================
# 5.1 Modelo supervisado: Random Forest Regressor
# ============================================================

from pyspark.ml.regression import RandomForestRegressor
from pyspark.ml.evaluation import RegressionEvaluator

# División de datos: 80% entrenamiento, 20% prueba.
train_df, test_df = df_preparado.randomSplit([0.8, 0.2], seed=42)

print("Registros de entrenamiento:", train_df.count())
print("Registros de prueba:", test_df.count())

# Modelo base.
rf = RandomForestRegressor(
    featuresCol="features",
    labelCol="PROMEDIO",
    predictionCol="prediction",
    numTrees=50,
    maxDepth=5,
    seed=42
)

modelo_rf = rf.fit(train_df)
predicciones_rf = modelo_rf.transform(test_df)

# Muestra de predicciones.
predicciones_rf.select("PROMEDIO", "prediction").show(20, truncate=False)


## 6. Métricas del modelo supervisado

Para evaluar el modelo de regresión se usan tres métricas:

- **RMSE:** mide el error promedio del modelo penalizando más los errores grandes. Mientras más bajo, mejor.
- **MAE:** mide el error absoluto promedio. Es más fácil de interpretar porque está en la misma escala del puntaje.
- **R²:** indica qué proporción de la variabilidad del puntaje es explicada por el modelo. Mientras más cerca esté de 1, mejor.


In [ ]:
# ============================================================
# 6.1 Métricas del Random Forest
# ============================================================

evaluador_rmse = RegressionEvaluator(
    labelCol="PROMEDIO",
    predictionCol="prediction",
    metricName="rmse"
)

evaluador_mae = RegressionEvaluator(
    labelCol="PROMEDIO",
    predictionCol="prediction",
    metricName="mae"
)

evaluador_r2 = RegressionEvaluator(
    labelCol="PROMEDIO",
    predictionCol="prediction",
    metricName="r2"
)

rmse_rf = evaluador_rmse.evaluate(predicciones_rf)
mae_rf = evaluador_mae.evaluate(predicciones_rf)
r2_rf = evaluador_r2.evaluate(predicciones_rf)

print("Métricas Random Forest base")
print("RMSE:", rmse_rf)
print("MAE:", mae_rf)
print("R²:", r2_rf)


## 6.2 Pruebas con diferentes parámetros para Random Forest

El enunciado solicita probar diferentes parámetros para cada técnica.  
Para Random Forest se prueban diferentes combinaciones de:

- `numTrees`: número de árboles del bosque.
- `maxDepth`: profundidad máxima de cada árbol.

El mejor modelo será aquel que reduzca el RMSE y MAE, y aumente el R².


In [ ]:
# ============================================================
# 6.2 Pruebas de parámetros: Random Forest
# ============================================================

resultados_rf = []

for num_trees in [20, 50, 100]:
    for max_depth in [3, 5, 8]:

        rf_temp = RandomForestRegressor(
            featuresCol="features",
            labelCol="PROMEDIO",
            predictionCol="prediction",
            numTrees=num_trees,
            maxDepth=max_depth,
            seed=42
        )

        modelo_temp = rf_temp.fit(train_df)
        pred_temp = modelo_temp.transform(test_df)

        rmse_temp = evaluador_rmse.evaluate(pred_temp)
        mae_temp = evaluador_mae.evaluate(pred_temp)
        r2_temp = evaluador_r2.evaluate(pred_temp)

        resultados_rf.append(
            (num_trees, max_depth, float(rmse_temp), float(mae_temp), float(r2_temp))
        )

resultados_rf_df = sparkMolina.createDataFrame(
    resultados_rf,
    ["numTrees", "maxDepth", "RMSE", "MAE", "R2"]
)

resultados_rf_df.orderBy("RMSE").show(truncate=False)


## 5.2 Modelo no supervisado: K-Means

K-Means permite identificar grupos de registros con características similares.  
En este proyecto, los clusters pueden interpretarse como perfiles de estudiantes o municipios según conectividad, condiciones familiares, inversión educativa, riesgo territorial y rendimiento académico.


In [ ]:
# ============================================================
# 5.2 Modelo no supervisado: K-Means
# ============================================================

from pyspark.ml.clustering import KMeans
from pyspark.ml.evaluation import ClusteringEvaluator

# Modelo base con k=3.
kmeans = KMeans(
    featuresCol="features",
    predictionCol="cluster",
    k=3,
    seed=42
)

modelo_kmeans = kmeans.fit(df_preparado)
clusters = modelo_kmeans.transform(df_preparado)

clusters.select(
    "PROMEDIO",
    "INTERNET_PER_CAPITA_1000",
    "cluster"
).show(20, truncate=False)


## 6.3 Métrica del modelo no supervisado: Silhouette

Para evaluar K-Means se usa la métrica **Silhouette**.

Esta métrica indica qué tan bien separados están los clusters:

- Valores cercanos a **1** indican clusters bien definidos.
- Valores cercanos a **0** indican clusters superpuestos.
- Valores negativos indican que algunos registros podrían estar asignados al cluster incorrecto.


In [ ]:
# ============================================================
# 6.3 Métrica Silhouette para K-Means
# ============================================================

evaluador_cluster = ClusteringEvaluator(
    featuresCol="features",
    predictionCol="cluster",
    metricName="silhouette"
)

silhouette_kmeans = evaluador_cluster.evaluate(clusters)

print("Silhouette KMeans base:", silhouette_kmeans)


## 6.4 Pruebas con diferentes valores de K

Para K-Means se prueban diferentes valores de `k`, que representa el número de grupos.  
El mejor valor de `k` será aquel que tenga mejor Silhouette y que además sea interpretable desde el negocio.


In [ ]:
# ============================================================
# 6.4 Pruebas de parámetros: K-Means
# ============================================================

resultados_kmeans = []

for k in [2, 3, 4, 5, 6]:

    kmeans_temp = KMeans(
        featuresCol="features",
        predictionCol="cluster",
        k=k,
        seed=42
    )

    modelo_temp = kmeans_temp.fit(df_preparado)
    clusters_temp = modelo_temp.transform(df_preparado)

    silhouette_temp = evaluador_cluster.evaluate(clusters_temp)

    resultados_kmeans.append((k, float(silhouette_temp)))

resultados_kmeans_df = sparkMolina.createDataFrame(
    resultados_kmeans,
    ["k", "Silhouette"]
)

resultados_kmeans_df.orderBy(col("Silhouette").desc()).show(truncate=False)


## 6.5 Interpretación de clusters

Después de elegir el mejor valor de `k`, se revisa el promedio de las principales variables por cluster.  
Esto permite interpretar cada grupo y relacionarlo con posibles acciones del Ministerio de Educación.


In [ ]:
# ============================================================
# 6.5 Interpretación de clusters
# ============================================================

from pyspark.sql.functions import avg, count

# Elegimos k=3 como modelo base interpretativo.
# Si la tabla anterior muestra que otro k tiene mejor Silhouette,
# se puede cambiar este valor.
k_optimo = 3

kmeans_final = KMeans(
    featuresCol="features",
    predictionCol="cluster",
    k=k_optimo,
    seed=42
)

modelo_kmeans_final = kmeans_final.fit(df_preparado)
clusters_final = modelo_kmeans_final.transform(df_preparado)

# Variables disponibles para interpretar clusters.
variables_interpretacion = [
    c for c in [
        "PROMEDIO",
        "INTERNET_PER_CAPITA_1000",
        "NO DE ACCESOS",
        "POBLACION_MUNICIPIO",
        "SGP_CALIDAD_PRESUPUESTO",
        "Exposición a Sequia Extrema"
    ]
    if c in clusters_final.columns
]

expresiones_agg = [
    avg(c).alias(f"promedio_{c.replace(' ', '_')}")
    for c in variables_interpretacion
]

clusters_final.groupBy("cluster").agg(
    count("*").alias("cantidad_registros"),
    *expresiones_agg
).orderBy("cluster").show(truncate=False)


# Análisis de resultados

El modelo supervisado Random Forest permite estimar el rendimiento promedio ICFES usando variables de conectividad, condiciones del hogar, contexto territorial e inversión educativa. Las métricas RMSE y MAE indican el tamaño promedio del error de predicción, mientras que R² permite analizar qué tanto logra explicar el modelo sobre la variabilidad del rendimiento.

Si el RMSE y el MAE son bajos en comparación con la escala del puntaje, el modelo tiene un comportamiento aceptable. Si el R² es cercano a 1, el modelo explica una parte importante del rendimiento académico; si es cercano a 0, significa que las variables usadas no son suficientes para explicar completamente el puntaje, lo cual también es un hallazgo importante porque sugiere que el desempeño académico depende de factores adicionales.

El modelo no supervisado K-Means permite segmentar los registros en grupos con características similares. Esta segmentación es útil para el Ministerio porque permite identificar perfiles de intervención. Por ejemplo, puede haber clusters con baja conectividad y bajo rendimiento, clusters con mejor conectividad y mejor desempeño, o clusters con condiciones mixtas donde el acceso a internet existe pero no necesariamente se traduce en mejores resultados académicos.


# Conclusión completa

A partir de la aplicación de técnicas de aprendizaje de máquina con Spark MLlib, se evidencia que el rendimiento académico medido mediante el promedio de los módulos ICFES puede analizarse desde una perspectiva integral que combine variables educativas, tecnológicas, territoriales y socioeconómicas.

El modelo supervisado Random Forest Regressor permitió construir una aproximación predictiva del puntaje promedio. Las métricas RMSE, MAE y R² permiten evaluar qué tan cercanas son las predicciones frente a los valores reales y qué tanta variabilidad del rendimiento logra explicar el modelo. La comparación de diferentes parámetros permite seleccionar una configuración más adecuada y evita depender de un único experimento.

Por otra parte, el modelo no supervisado K-Means permitió agrupar los registros en perfiles similares. Esta segmentación resulta útil para la toma de decisiones públicas, porque no todos los municipios o estudiantes requieren el mismo tipo de intervención. Algunos grupos pueden requerir inversión prioritaria en conectividad, otros pueden necesitar fortalecimiento institucional, y otros pueden requerir estrategias complementarias asociadas a condiciones del hogar o vulnerabilidad territorial.

En conclusión, los modelos de MLlib enriquecen el análisis descriptivo realizado previamente porque permiten pasar de observar relaciones generales a construir herramientas de predicción y segmentación. Para el Ministerio de Educación, estos resultados pueden apoyar la priorización de acciones territoriales, especialmente en municipios donde coinciden baja conectividad, menor rendimiento académico y mayores condiciones de vulnerabilidad.
